### ISSUE TITLE
Some image add device successfully, but the device file seems strange.

### ISSUE THREAD

**Issue by qq690388648** [lxc/lxd#3121]
``` shell
lxc info 
  driver: lxc
  driverversion: 2.0.7
  kernel: Linux
  kernelarchitecture: x86_64
  kernelversion: 3.10.0-327.44.58.28_6.x86_64
  server: lxd
  serverpid: 11936
  serverversion: 2.0.9
  storage: dir

```

Run the command below:
`lxc config device add test-centos3 sdb unix-block path=/dev/sdb`
It returns:
`Device sdb added to test-centos3`
But when I list it in the container,I find it is not a block file, so it can not be mounted.
``` shell
ls -l /dev/sdb
-rwxr-xr-x 1 root root 0 Mar 27 03:46 /dev/sdb
```
**PS:it just found in some images(centos for me), but work well in other images(ubuntu for me).**

**Comment by stgraber** [comment]
The way LXD sets up those devices is:
 - Touch them as an empty file
 - Bind-mount the device file on top

So the fact that it shows up as just an empty file in your case suggests that something unmounted it.

**Comment by qq690388648** [comment]
I have read the code of lxd and lxc, and find some difference between their way of mounting.
In lxd, the steps are(use the example in issue):
* Mknod under `/var/lib/lxd/devices/test-centos3/unix.dev-sdb`
* Bind-mount the `unix.dev-sdb` to `/var/lib/lxd/shmounts/test-centos3/lxdmount_298706992`(Which has been mounted in `/dev/.lxd-mounts` inner container).
* Change user namespace and mount namespace to namespace of container.
* Touch an empty file `/dev/sdb` inner container.
* Move-mount to `/dev/sdb` inner container.

But when container is restarted, the lxd write the config devices into `lxc.conf` as `lxc.mount.entry = /var/lib/lxd/devices/test-centos3/unix.dev-sdb dev/sdb none bind,create=file`, and lxc deal with it in the steps below:

* Touch an empty file `/var/lib/lxd/container/rootfs/dev/sdb`.
* Bind-mount the `unix.dev-sdb` to `/var/lib/lxd/container/rootfs/dev/sdb`.

In my case, the `/dev/sdb` look strange when I run command `lxc config device add test-centos3 sdb unix-block path=/dev/sdb`, but when I restart the container by `lxc restart test-centos3`, the device file get into right mode.
``` shell
ls -l /dev/sdb 
brw-rw---- 1 root root 8, 16 Mar 27 07:44 /dev/sdb
```

But I really cann't find what cause this.

**Comment by stgraber** [comment]
What exact container images are you using?

**Comment by qq690388648** [comment]
An image made it by myself. I made it with the below steps.
* Got a kvm image with raw disk format.
* Mount it as a loop device.
* kpartx it into 2 part.
* Tar all files in part2 as rootfs.
* Touch a simplest metadata file.

**Comment by stgraber** [comment]
Sorry it took me so long to circle back to this bug. I'll attempt to reproduce it with a clean Centos 6 and Centos 7 containers on LXD 2.0.9. If I can't reproduce it, the only real assumption I'd be left with is that something in your custom image is unmounting the bind-mount.

I know systemd has been known to do such things in a limited number of cases, though I've never seen it myself. Looking at the systemd journal may give some hints as to what's going on, if that's the case.

**Comment by qq690388648** [comment]
ok, thanks for your patience!!!

**Comment by stgraber** [comment]
So I just tried with LXD 2.0.10 on Ubuntu 16.04 using the "dir" backend, so similar setup as the reporter.
I then did:
 - lxc launch images:centos/6 centos6
 - lxc launch images:centos/7 centos7
 - lxc config device add centos6 vda unix-block path=/dev/vda
 - lxc config device add centos7 vda unix-block path=/dev/vda

This gave me:
```
root@vm02:~# lxc exec centos6 -- ls -lh /dev/vda
brw-rw---- 1 root root 253, 0 Jun 26 00:55 /dev/vda
root@vm02:~# lxc exec centos7 -- ls -lh /dev/vda
brw-rw---- 1 root root 253, 0 Jun 26 04:55 /dev/vda
```

And just to confirm we get the same when we have the bind-mounts setup by liblxc on startup:
```
root@vm02:~# lxc restart centos6 centos7
root@vm02:~# lxc exec centos6 -- ls -lh /dev/vda
brw-rw---- 1 root root 253, 0 Jun 26 00:57 /dev/vda
root@vm02:~# lxc exec centos7 -- ls -lh /dev/vda
brw-rw---- 1 root root 253, 0 Jun 26 04:57 /dev/vda
```

So whatever the issue may be here, it doesn't appear to happen on the clean images, suggesting that it's something going on inside the container which may cause the reported issue.

Closing the issue for now as there's no evidence that LXD is the source of the problem here.
